src/main/java/org/threadly/concurrent/PriorityScheduler.java

```diff
@@ -1217,94 +1140,107 @@ protected void shutdownService() {
+            // no tasks, so just wait till any tasks come in
+            LockSupport.park();
+          } else {
+            // only high priority task is queued, so run if ready
+            long nextTaskDelay = nextHighPriorityTask.getDelay(TimeUnit.MILLISECONDS);
+            if (nextTaskDelay <= 0) {
+              if (nextHighPriorityTask.canExecute()) {
+                return nextHighPriorityTask;
               }
-            } else if (executeQueue.remove(nextExecuteTask)) {
-              // if we can remove the task (aka it has not been removed already), we can execute it
-              nextExecuteTask.executing();
-              return nextExecuteTask;
+            } else {
+              LockSupport.parkNanos(Clock.NANOS_IN_MILLISECOND * nextTaskDelay);
             }
-          } else if (executeQueue.remove(nextExecuteTask)) {
-            // if we can remove the task (aka it has not been removed already), we can execute it
-            nextExecuteTask.executing();
-            return nextExecuteTask;
           }
-        } else if (nextScheduledTask != null) {
-          if (nextScheduledTask.getDelay(TimeUnit.MILLISECONDS) <= 0) {
-            synchronized (scheduleQueue.getModificationLock()) {
-              // scheduled tasks must be removed, and call .executing() while holding the lock
-              if (scheduleQueue.remove(nextScheduledTask)) {
-                nextScheduledTask.executing();
-                return nextScheduledTask;
-              }
+        } else if (nextHighPriorityTask == null) {
+          // only low priority task is queued, so run if ready
+          long nextTaskDelay = nextLowPriorityTask.getDelay(TimeUnit.MILLISECONDS);
+          if (nextTaskDelay <= 0) {
+            if (nextLowPriorityTask.canExecute()) {
+              return nextLowPriorityTask;
             }
           } else {
-            workerPool.handleEstimatedTimeTillNextTask(this, nextScheduledTask.getDelayEstimateInMs());
-            LockSupport.parkNanos(Clock.NANOS_IN_MILLISECOND * nextScheduledTask.getDelay(TimeUnit.MILLISECONDS));
+            LockSupport.parkNanos(Clock.NANOS_IN_MILLISECOND * nextTaskDelay);
           }
         } else {
-          workerPool.handleEstimatedTimeTillNextTask(this, Long.MAX_VALUE);
-          LockSupport.park();
+          // both tasks are available, so see which one makes sense to run
+          long nextHighDelay;
+          long nextLowDelay;
+          ClockWrapper.stopForcingUpdate();
+          try {
+            nextHighDelay = nextHighPriorityTask.getDelay(TimeUnit.MILLISECONDS);
+            nextLowDelay = nextLowPriorityTask.getDelay(TimeUnit.MILLISECONDS);
+          } finally {
+            ClockWrapper.resumeForcingUpdate();
+          }
+          if (nextHighDelay <= 0) {
+            // high task is ready, make sure it should be picked
+            if (nextHighDelay <= nextLowDelay || nextLowDelay > maxWaitForLowPriorityInMs * -1) {
```

Comment:
Here was the intention of this logic:

* If there is ever a high task that is delayed longer than a low task, execute the high task
* If there is a high task and a low task ready to run, and the above condition is not true, run the high task unless the low task has been waiting it's maximum wait time
* If there is just a high or just a low task ready to run, run it immediately

So the second condition is doing the second bullet, at that point we have both a high and a low priority task. But if the low priority task has not been waiting it's maximum wait time, we want to favor the high priority task.

Do you think it makes sense now?